beginner_source/knowledge_distillation_tutorial.py

```diff
@@ -0,0 +1,651 @@
+
+######################################################################
+#We need to modify our test function for the same reason. Here we ignore the hidden representation returned by the model.
+
+def test_multiple_outputs(model, test_loader, device):
+    model.to(device)
+    model.eval()
+
+    correct = 0
+    total = 0
+
+    with torch.no_grad():
+        for inputs, labels in test_loader:
+            inputs, labels = inputs.to(device), labels.to(device)
+
+            outputs, _ = model(inputs) # In this line we disregard the second tensor of the tuple
+            _, predicted = torch.max(outputs.data, 1)
+
+            total += labels.size(0)
+            correct += (predicted == labels).sum().item()
+
+    accuracy = 100 * correct / total
+    print(f"Test Accuracy: {accuracy:.2f}%")
+    return accuracy
+
+######################################################################
+# In this case we could easily include both knowledge distillation and cosine loss minimization in the same function. It is common to combine methods to achieve better performance in teacher-student paradigms.
+# For now we can run a simple train-test session.
+
+#Train and test the lightweight network with cross entropy loss
+train_cosine_loss(teacher=modified_nn_deep, student=modified_light_nn, train_loader=train_loader, epochs=10, learning_rate=0.001, hidden_rep_loss_weight=0.25, ce_loss_weight=0.75, device=device)
+test_accuracy_light_ce_and_cosine_loss = test_multiple_outputs(modified_light_nn, test_loader, device)
+
+######################################################################
+# Our naive minimization does not guarantee better results for a number of reasons, one being the dimensionality of the vectors. Cosine similarity in general works better than Euclidean distance for vectors of larger dimensionality, but we were dealing with vectors with ``1024`` components each, so it is a lot harder to extract meaningful similarities.
+# Furthermore as we mentioned, pushing towards a match of the hidden representation of the teacher and the student is not supported by theory. There are no good reasons why we should be aiming for an 1-1 match of these vectors.
+# We will provide a final example of training intervention, by including an extra network called regressor. Our intention is to extract the feature map of the teacher after a convolutional layer, extract a feature map of the student after a convolutional layer and try to match these maps, but this time include a regressor between the networks.
+# The regressor will be trainable and ideally will do a better job than our naive cosine loss minimization scheme. Its main job is to match the dimensionality of these feature maps so that we can properly define a loss function between the teacher and the student. Defining such loss function provides a teaching "path", which is basically a flow to back-propagate gradients, that will change the student's weights.
```

Comment:
```suggestion
# The regressor will be trainable and ideally will do a better job than our naive cosine loss minimization scheme. Its main job is to match the dimensionality of these feature maps so that we can properly define a loss function between the teacher and the student. Defining such a loss function provides a teaching "path," which is basically a flow to back-propagate gradients that will change the student's weights.
```